### PR TITLE
Add ja3 signature to handshake session

### PR DESCRIPTION
**Pull request by ajathan**

### CODE REVIEW

File: src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java [hunk 1]
```diff
@@ -39,30 +47,35 @@ public SSLEngineResult wrap(final ByteBuffer[] srcs, final int offset, final int
...
+			//unwrap
+			result = engine.unwrap(src, dsts, offset, length);
+			
+			final SSLSession handshakeSession = engine.getHandshakeSession();
+			if (handshakeSession != null) {
+				//set ja3 signature in handshake session
+				handshakeSession.putValue(JA3Constants.JA3_FINGERPRINT, ja3);
+				isJa3SignatureSet = true;
+			}
+		} else {
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java]: I haven't fully reviewed this code. The idea is to make the flows to unwrap post handshake fast with as many less checks as possible. The handshake + ja3 generation can be somewhat expensive and still efficient. Can you review the code and see?

**Comment by nvsnvikram** [comment]
👎 

Can you also add README.md to this project?

File: src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java [hunk 2]
```diff
@@ -20,7 +20,15 @@
      */
     private final SSLEngine engine;
 
-    private boolean generatedJa3Signature = false;
+    /**
+     * True if the ja3 signature has been set in the ssl session.
+     */
+    private boolean isJa3SignatureSet = false;
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java]: How is this different from ja3 != null?

Comment (nvsnvikram) [reply src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java]: ok, see it now. This is ja3 completed.

File: src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java [hunk 3]
```diff
@@ -39,30 +47,35 @@ public SSLEngineResult wrap(final ByteBuffer[] srcs, final int offset, final int
...
-            if (ja3 != null) {
-                // no lock needed, it is fine if we write same digest multiple times
-                engine.getSession().putValue(JA3Constants.JA3_FINGERPRINT, ja3);
-                generatedJa3Signature = true;
-            }
-        }
-
-
-        return result;
+		final SSLEngineResult result;
```
Comment (nvsnvikram) [inline src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java]: Would this work?

    @Override
    public SSLEngineResult unwrap(final ByteBuffer src, final ByteBuffer[] dsts, final int offset, final int length) throws SSLException {
		if (!ja3Done) {
		    if (ja3Signature != null) {
		        final SSLSession handshakeSession = engine.getHandshakeSession();
		        if (handshakeSession != null) {
		            // Set ja3 signature in handshake session
		            handshakeSession.putValue(JA3Constants.JA3_FINGERPRINT, ja3);
		            ja3Done = true;
		        }
		    } else {
		        // 1. Generate JA3 signature
		        final HandshakeStatus handshakeStatus = engine.getHandshakeStatus();
		        if (HandshakeStatus.FINISHED == handshakeStatus) {
				    ja3Done = true;
			    } else {
			        ja3 = new JA3Signature().ja3Signature(src);
			    }
		    }
		}

		return engine.unwrap(src, dsts, offset, length);
    }

**Review by nvsnvikram — CHANGES_REQUESTED** [review]
Please review my changes and reply back

File: src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java (hunk 3 above)
Comment (ajathan) [reply src/main/java/com/lafaspot/ja3_4java/JA3SSLEngineWrapper.java]: done

**Review by nvsnvikram — APPROVED** [review]

**Pull request merged**